poet/poet.py

```diff
@@ -76,10 +85,106 @@ def walk(package):
+    name: str
+    homepage: str
+    url: str
+    checksum: str
+    checksum_type: str
+
+def get_download_url_from_pip_source_file(module: str, pip_source_file: Path, output_dir=os.getenv("PIP_SOURCE_DIR")) -> str:
+    """
+    Returns the download URL for the pip source distribution.
+    This method will download the pip package from the source distribution. 
+
+    The standard out of this command contains an obfuscated URL and a regular URL that points to a .tar.gz file.
+
+    Args:
+        module (str): The name of the module to download.
+        output_dir (str): The directory to download the module to.
+
+    Returns:
+        str: The download URL for the pip source distribution.
+    """
+    try:
+        output = subprocess.run(shlex.split(f"pip download --dest {output_dir} --no-binary :all: --no-deps {module}"), capture_output=True, text=True)
+    except subprocess.CalledProcessError as cpe:
+        raise PipSourceMetadataException(f"Could not download {module} from pip source file: {cpe.stderr}")
+    
+    try:
+        extractor = URLExtract()
+        urls = extractor.find_urls(output.stdout)
+        return [url for url in urls if pip_source_file.name in url][0]
+    except Exception as e:
+        raise PipSourceMetadataException(f"Could not get download URL from pip source file: {e}") from e
+
+def get_checksum_from_pip_source_file(pip_source_file: Path) -> str:
+    """Given the path to a pip source file, return the files checksum.
+
+    Args:
+        pip_source_file (Path): The path to a .tar.gz file containing a pip source distribution.
+
+    Returns:
+        str: The checksum of the pip source file.
+    """
+    if not pip_source_file.exists():
+        raise PipSourceMetadataException("File does not exist: %s" % pip_source_file)
+    
+    return sha256(pip_source_file.read_bytes()).hexdigest()
+
+
+def get_metadata_from_pip_source(package_name: str, pip_source_file: Path) -> PackageMetadata:
+    """Given the path to a pip source file, return a PackageMetadata object.
+
+    Args:
+        pip_source_file (Path): The path to a .tar.gz file containing a pip source distribution.
+
+    Returns:
+        PackageMetadata: A dictionary of metadata about the package required for the resource stanza.
+    """
+    if not pip_source_file.exists():
+        raise PipSourceMetadataException("File does not exist: %s" % pip_source_file)
+    
+    try:
+        metadata_object = metadata(pip_source_file)
+    except Exception as e:
+        raise PipSourceMetadataException("Could not get metadata from pip source file: %s" % e)
+
+    return PackageMetadata(
+        name=metadata_object.get("Name"),
+        homepage=metadata_object.get("Home-page"),
+        url=get_download_url_from_pip_source_file(package_name, pip_source_file),
+        checksum=get_checksum_from_pip_source_file(pip_source_file),
+        checksum_type="sha256"
+    )
+
+def research_package(name: str, version=None) -> PackageMetadata:
```

Comment:
if you want `PackageMetadata` to be API-compatible with the `dict` that this function used to return, then `PackageMetadata` needs to be indexable:
```python
pkg = research_package("indigoag.devops.common")
pkg["name"]  # i expect this will throw a TypeError with your current implementation
```
to make `PackageMetadata` indexable, implement `__getitem__()` and `__setitem__()` on it, like so:
```python
@dataclass 
class PackageMetadata:
    name: str
    ...

    def __getitem__(self, idx):
        try:
            return getattr(self, idx)
        except AttributeError as err:
            raise KeyError(idx) from err

    def __setitem__(self, idx, val):
        setattr(self, idx, val)
```

Reply:
Addressed in: [b8abb61](https://github.com/indigo-ag/homebrew-pypi-poet/pull/1/commits/b8abb6137f1b6d66ddc05e52473cc3c83aec98a9)